### ISSUE TITLE
Debugger: local variables

### ISSUE THREAD

**Issue by akharche** [IntelPython/numba-dppy#338]
Implement `info locals` for all the kernel arguments and local variables.

Expected:

- Debugger shows all the local variables **names** inside the numba-dppy kernel when a user calls `info locals`.
- Debugger shows the **values** of these local variables when a user calls `info locals` or `print variable`.
- Debugger shows a **type** of the local variables `ptype variable`.
- User can reset local variable value.

Current state and issues:

1. Some kernel arguments are optimized out when we set optimization level to 1 (by default). Disabling optimization ("-O0") causes segfault due to unsafe address space cast. Solution is in progress: #364 
2. Local variables created inside a kernel don't have a LLVM debug tag #320. Temporary workaround was made to Numba (https://github.com/IntelPython/numba/pull/217). TODO: update results with Numba 0.53.
3. For some reason GDB with Numba shows variable values only on the second entry to the compiled function. #382

**Comment by akharche** [comment]
Numba `info locals`:

```
@njit(debug=True)
def foo(arg):
    l1 = arg + 6        	# numba-kernel-breakpoint
    l2 = arg * 5.43             # second-line
    l3 = (arg, l1, l2, "bar")   # third-line
    print(arg, l1, l2, l3)     # fourth-line
```

gdb output for Numba-jit function:
![image](https://user-images.githubusercontent.com/41467010/115873106-3abe8900-a44b-11eb-8cd9-c9911b76a852.png)

Printing arguments of jit-function is confusing, they are represented as IR lowered values: 
![image](https://user-images.githubusercontent.com/41467010/115873924-30e95580-a44c-11eb-85df-29654b2f1baf.png)

**Comment by akharche** [comment]
Numba "info locals" returns local values in not readable view. It's impossible to understand the value of arguments

![image](https://user-images.githubusercontent.com/41467010/115874310-9ccbbe00-a44c-11eb-9e33-b0d0bbc414c2.png)

**Comment by PokhodenkoSA** [comment]
Strange behavior:
`foo(arg)` called twice: with `arg = 0` and with `arg = 987`. It is unexpected for user. But Numba docs contain [example](https://numba.pydata.org/numba-doc/latest/user/troubleshoot.html?highlight=numba%20gdb#example-debug-usage) with same behavior.

**Comment by akharche** [comment]
Forcing Numba to emit debug info for all the variables ([pr](https://github.com/IntelPython/numba/pull/217/files)) gives the following result:

```
@dppy.kernel
def data_parallel_sum(user_a, user_b, user_c):
    i = dppy.get_global_id(0)
    user_d = 10
    user_c[i] = user_a[i] + user_b[i] + user_d

def driver(a, b, c, global_size):
    data_parallel_sum[global_size, dppy.DEFAULT_LOCAL_SIZE](a, b, c)

def main():
    global_size = 10
    N = global_size
    a = np.array(np.random.random(N), dtype=np.float32)
    b = np.array(np.random.random(N), dtype=np.float32)
    c = np.ones_like(a)
    with dpctl.device_context("opencl:gpu") as gpu_queue:
        driver(a, b, c, global_size)

```
![image](https://user-images.githubusercontent.com/41467010/117980656-dbaeae80-b33c-11eb-9759-50c63510657a.png)

**Comment by akharche** [comment]
Currently we see all the user variables names while debugging, but they don't have any value.

**Comment by akharche** [comment]
With Numba 0.53 local variables created inside the kernel are shown:

![image](https://user-images.githubusercontent.com/41467010/120632796-09859f80-c472-11eb-96cf-8511b84861f6.png)

**Comment by PokhodenkoSA** [comment]
Fixed in Numba 0.55.
